src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/GeneratePortNumberConfig.cs

```diff
@@ -0,0 +1,145 @@
+{
+    internal class GeneratePortNumberConfig : BaseMacroConfig<GeneratePortNumberMacro, GeneratePortNumberConfig>
+    {
+        private const int LowPortDefault = 1024;
+        private const int HighPortDefault = 65535;
+
+        private static readonly object LockObj = new();
+
+        private static readonly HashSet<int> UnsafePorts = new()
+        {
+                    2049, // nfs
+                    3659, // apple-sasl / PasswordServer
+                    4045, // lockd
+                    6000, // X11
+                    6665, // Alternate IRC [Apple addition]
+                    6666, // Alternate IRC [Apple addition]
+                    6667, // Standard IRC [Apple addition]
+                    6668, // Alternate IRC [Apple addition]
+                    6669, // Alternate IRC [Apple addition]
+        };
+
+        private static readonly HashSet<int> AllocatedPorts = new();
+
+        internal GeneratePortNumberConfig(GeneratePortNumberMacro macro, string variableName, string? dataType, int fallback, int low, int high)
+             : base(macro, variableName, dataType)
+        {
+            Port = AllocatePort(low, high, fallback);
+        }
+
+        internal GeneratePortNumberConfig(GeneratePortNumberMacro macro, IGeneratedSymbolConfig generatedSymbolConfig)
+            : base(macro, generatedSymbolConfig.VariableName, generatedSymbolConfig.DataType)
+        {
+            int low = GetOptionalParameterValue(generatedSymbolConfig, nameof(Low), ConvertJTokenToInt, LowPortDefault);
+            int high = GetOptionalParameterValue(generatedSymbolConfig, nameof(High), ConvertJTokenToInt, HighPortDefault);
+            if (low < LowPortDefault)
+            {
+                low = LowPortDefault;
+            }
+
+            if (high > HighPortDefault)
+            {
+                high = HighPortDefault;
+            }
+
+            if (low > high)
+            {
+                low = LowPortDefault;
+                high = HighPortDefault;
+            }
+
+            int fallback = GetOptionalParameterValue(generatedSymbolConfig, "fallback", ConvertJTokenToInt, 0);
+
+            Port = AllocatePort(low, high, fallback);
+        }
+
+        internal int Port { get; }
+
+        internal int Low { get; }
+
+        internal int High { get; }
+
+        private static int AllocatePort(int low, int high, int fallback = 0)
+        {
+            int startPort = CryptoRandom.NextInt(low, high);
+
+            for (int testPort = startPort; testPort <= high; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
+
+            for (int testPort = low; testPort < startPort; testPort++)
+            {
+                if (TryAllocatePort(testPort, out int port))
+                {
+                    return port;
+                }
+            }
```

Comment:
The two cycles can get consolidated into single one:

```
for (int testPort = low; testPort < high; testPort++)
{
      if (TryAllocatePort((testPort + startPort) % high, out int port))
      {
            return port;
      }
}
```

just not 100% if it's more readable...

Reply:
I don't think it works like this or i didn't get it:

given low = 300, high = 500, startPort = 400
the first port to be allocated is (300 + 400) % 500 = 200, which is out of the range.

likely it should be 
```
(testPort + startPort - low) % high
```

i suggest we keep it as is :)